internal-api/src/main/java/datadog/trace/api/TagMap.java

```diff
@@ -0,0 +1,2208 @@
+package datadog.trace.api;
+
+import datadog.trace.api.function.TriConsumer;
+import java.util.AbstractCollection;
+import java.util.AbstractSet;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * A super simple hash map designed for...
+ *
+ * <ul>
+ *   <li>fast copy from one map to another
+ *   <li>compatibility with Builder idioms
+ *   <li>building small maps as fast as possible
+ *   <li>storing primitives without boxing
+ *   <li>minimal memory footprint
+ * </ul>
+ *
+ * <p>This is mainly accomplished by using immutable entry objects that can reference an object or a
+ * primitive. By using immutable entries, the entry objects can be shared between builders & maps
+ * freely.
+ *
+ * <p>This map lacks some features of a regular java.util.Map...
+ *
+ * <ul>
+ *   <li>Entry object mutation
+ *   <li>size tracking - falls back to computeSize
+ *   <li>manipulating Map through the entrySet() or values()
+ * </ul>
+ *
+ * <p>Also lacks features designed for handling large maps...
+ *
+ * <ul>
+ *   <li>bucket array expansion
+ *   <li>adaptive collision
+ * </ul>
+ */
+
+/*
+ * For memory efficiency, TagMap uses a rather complicated bucket system.
+ * <p>
+ * When there is only a single Entry in a particular bucket, the Entry is stored into the bucket directly.
+ * <p>
+ * Because the Entry objects can be shared between multiple TagMaps, the Entry objects cannot contain
+ * form a link list to handle collisions.
+ * <p>
+ * Instead when multiple entries collide in the same bucket, a BucketGroup is formed to hold multiple entries.
+ * But a BucketGroup is only formed when a collision occurs to keep allocation low in the common case of no collisions.
+ * <p>
+ * For efficiency, BucketGroups are a fixed size, so when a BucketGroup fills up another BucketGroup is formed
+ * to hold the additional Entry-s.  And the BucketGroup-s are connected via a linked list instead of the Entry-s.
+ * <p>
+ * This does introduce some inefficiencies when Entry-s are removed.
+ * In the current system, given that removals are rare, BucketGroups are never consolidated.
+ * However as a precaution if a BucketGroup becomes completely empty, then that BucketGroup will be
+ * removed from the collision chain.
+ */
+public final class TagMap implements Map<String, Object>, Iterable<TagMap.Entry> {
+  /** Immutable empty TagMap - similar to {@link Collections#emptyMap()} */
+  public static final TagMap EMPTY = createEmpty();
+
+  private static final TagMap createEmpty() {
+    return new TagMap().freeze();
```

Comment:
could we use the parameterized ctor to avoid allocating an array of 16 entries that'll never be used ?
```suggestion
    return new TagMap(null).freeze();
```
(not sure if null is OK, I haven't read all the code yet :p)

Reply:
so, yeah, not ok, but we could create an empty array.